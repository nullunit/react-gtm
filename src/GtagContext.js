import ArgumentError from './ArgumentError';

/**
 * {@link GtagContext} global site tag context singleton.
 * @type {GtagContext}
 */
let GTAG_CONTEXT = null;

/**
 * The Google global site tag context, allowing access to push events
 * to <a href="https://support.google.com/analytics/answer/1008015?hl=en">Google Analytics</a>
 * and/or <a href="https://support.google.com/tagmanager/answer/6102821?hl=en">Google Tag manager</a>.
 */
export default class GtagContext {

    constructor(window, id) {

        if (window == null) {
            throw new ArgumentError('GtagContext could not be created, required parameter \'window\' cannot be null');
        }

        if (id == null) {
            throw new ArgumentError('GtagContext could not be created, required parameter \'id\' cannot be null');
        }

        this._window = window;
        this._id = id;
        this._userId = null;
        this._clientId = null;
        this._appName = null;
        this._appVersion = null;
        this._dataLayer = this._window.dataLayer || [];
        this._window.dataLayer = this._dataLayer;
    }

    get dataLayer() {
        return this._dataLayer;
    }

    get id() {
        return this._id;
    }

    get appName() {
        return this._appName;
    }

    set appName(value) {

        if (this._appName === value) {
            return;
        }

        this._appName = value;
        this.push({ 'appName': value });
    }

    get appVersion() {
        return this._appVersion;
    }

    set appVersion(value) {

        if (this._appVersion === value) {
            return;
        }

        this._appVersion = value;
        this.push({ 'appVersion': value });
    }

    get clientId() {
        return this._clientId;
    }

    set clientId(value) {

        if (this._clientId === value) {
            return;
        }

        this._clientId = value;
        this.push({ 'client_id': value });
    }

    get userId() {
        return this._userId;
    }

    set userId(value) {

        if (this._userId === value) {
            return;
        }

        this._userId = value;
        this.push({ 'userId': value });
    }

    gtag() {
        this.push(arguments);
    }

    event(event, params) {
        this.push({ event, ...params });
    }

    push() {
        this._dataLayer.push(...arguments);
    }

    toString() {
        const id = this.id ? `${this.id} ` : '';
        const userId = this.userId ? `'${this.userId}'` : '<null>';
        const clientId = this.clientId ? `'${this.clientId}'` : '<null>';

        return `Gtag{ ${id}userId:${userId}, clientId:${clientId} }`;
    }

}

/**
 * Get or create the Google {@link GtagContext}
 * @param {Window} window Required for first call to this function, the javascript window object.
 * @param {String} id Required for first call to this function, the Google Analytics or Tag Manager ID.
 * @returns {GtagContext} Returns the Google Tag Manager context.
 */
export const getGtagContext = (...args) => {

    if (GTAG_CONTEXT == null) {
        const [ w, id ] = args;
        GTAG_CONTEXT = new GtagContext(w || window, id);
    }

    return GTAG_CONTEXT;
};
