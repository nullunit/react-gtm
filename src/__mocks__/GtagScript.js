import React from 'react';

/**
 * Mock skips adding Google Analytics script to page as well as rendering child components.
 */
export default class GtagScript extends React.Component {

    render() {
        return <h1>mock render Google Analytics script to page</h1>;
    }

}
