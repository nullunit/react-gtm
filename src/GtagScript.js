import GtagContext from './GtagContext';
import PropTypes from 'prop-types';
import React from 'react';

export default class GtagScript extends React.Component {

    static propTypes = {
        'children': PropTypes.node.isRequired,
        'context': PropTypes.instanceOf(GtagContext).isRequired,
        'id': PropTypes.string.isRequired
    };

    componentDidMount() {

        const { id } = this.props;
        const gtagScript = document.createElement('script');
        gtagScript.async = true;
        gtagScript.src = `https://www.googletagmanager.com/gtag/js?id=${id}`;

        const [ firstScript ] = document.getElementsByTagName('script');
        firstScript.parentNode.insertBefore(gtagScript, firstScript);

        const { context } = this.props;
        context.gtag('js', new Date());
        context.gtag('config', id);
    }

    render() {
        const { children } = this.props;
        return React.Children.only(children);
    }

}
