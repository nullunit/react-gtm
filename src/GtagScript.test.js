import GtagContext from './GtagContext';
import GtagScript from './GtagScript';
import React from 'react';
import renderer from 'react-test-renderer';

const UA_ID = 'UA-123456';

describe('GtagScript tests', () => {

    it('Append Google Universal Analytics script to load Google Analytics', () => {

        const script = '<script type="text/javascript" src="/static/js/bundle.js"></script>';
        const expectedGtagScript = `<script src="https://www.googletagmanager.com/gtag/js?id=${UA_ID}"></script>${script}`;

        document.body.innerHTML =
            '<html>' +
            `  <body>${script}</body>` +
            '</html>';

        const context = new GtagContext(window, UA_ID);
        const props = {
            'id': UA_ID,
            context,
            'useTagManager': false
        };

        //test render snapshot
        const json = renderer
            .create(<GtagScript { ...props }><h1>mock child</h1></GtagScript>)
            .toJSON();

        expect(json).toMatchSnapshot();
        expect(expectedGtagScript).toEqual(document.body.innerHTML.trim());
    });

});
