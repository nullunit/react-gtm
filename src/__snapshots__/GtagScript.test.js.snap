// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GtagScript tests Append Google Universal Analytics script to load Google Analytics 1`] = `
<h1>
  mock child
</h1>
`;
